### PR TITLE
[doc] Venice Improvement Proposal -- Design Doc process

### PR DESCRIPTION
**Pull request by manujose0**
## [doc] Venice Design Documentation Process 

Based on Venice open source contributers sync, we aligned on a design document process inspired by Apache Pulsar. Please read the design doc guide for more details. 

## How was this PR tested?

Build Complete, Doc change no functional changes. 
## Does this PR introduce any user-facing changes?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure to explain your proposed changes and call out the behavior change.

### CODE REVIEW

File: docs/dev_guide/design_doc.md
Comment (FelixGV) [inline docs/dev_guide/design_doc.md:20]: s/PIP/VIP/

There is at least one more below. Let's do a search and replace.

File: docs/dev_guide/design_doc.md
Comment (FelixGV) [inline docs/dev_guide/design_doc.md:59]: I think that doesn't work so well, because then the VIP-XXX.md file won't get committed to the repo. I think all of the PRs should be merged (unless perhaps they are discarded out of hand due to not respecting the template or other basic issues...). The status of the PR does not need to align to the status of the VIP.

File: docs/dev_guide/design_doc.md
Comment (FelixGV) [inline docs/dev_guide/design_doc.md:17]: Although it may be a bit of a tongue twister, I prefer that we use "functionality" rather than "feature", since the latter is ambiguous in the context of machine learning, which a lot of Venice use cases fall into.

File: docs/proposals/VIP_TEMPLATE.md
Comment (FelixGV) [inline docs/proposals/VIP_TEMPLATE.md:90]: I don't think we should mention GDPR. That seems overly specific and nowadays every region has their own such thing.

File: docs/proposals/VIP_TEMPLATE.md
Comment (FelixGV) [inline docs/proposals/VIP_TEMPLATE.md:102]: "nation" seems like a chopped word. "Explanation", perhaps?

File: docs/dev_guide/design_doc.md
Comment (FelixGV) [inline docs/dev_guide/design_doc.md:55]: I think we could add something here to the effect that if a deep dive is required, it should be done in a public slack channel named after the VIP, e.g. `#vip-123`.

File: docs/proposals/VIP_TEMPLATE.md
Comment (FelixGV) [inline docs/proposals/VIP_TEMPLATE.md:8]: I would like to have a convention that MD files, and in particular VIP files, should adhere to a line width of max 80 characters. In rendered mode, markdown will make the text flow regardless of where the line breaks are, but for the sake of Pull Request commenting, I think it will be useful to limit lines to a short width as that will make it easier to target specific parts of a paragraph (since the granularity of GitHub PR comments is per-line).

I wonder if spotless (or other) can do this for us automatically... but regardless, we can have it as a convention.

**Review by FelixGV — COMMENTED** [review]
Thanks a lot for kicking this off @manujose0!! I left a few comments.

File: docs/dev_guide/design_doc.md
Comment (manujose0) [reply docs/dev_guide/design_doc.md:59]: So the proposal in the Pulsar  was to close the rejected PR, this is not same as deleting it. It will be around as a closed PR. Looking for all the rejected proposal -- search pr:is_closed label:vip. 

But I think we can do it differently. In the design_doc.md page (landing page) we can categorize them into ACCEPTED and REJECTED categories. The issue with this approach is that we have to maintain the bookkeeping. Closing the PR avoids that bookkeeping efforts. 

This is why I went with the merged/closed approach so that its light weight on the process side.

File: docs/dev_guide/design_doc.md
Comment (manujose0) [reply docs/dev_guide/design_doc.md:17]: That is a good point. s/feature/functionality applied.

File: docs/proposals/VIP_TEMPLATE.md
Comment (manujose0) [reply docs/proposals/VIP_TEMPLATE.md:90]: I thought GDPR is like a norm these days. These are indicative anyway, will remove.

File: docs/dev_guide/design_doc.md
Comment (manujose0) [reply docs/dev_guide/design_doc.md:55]: I was thinking of having our committer sync meeting to pick up these VIP as a topic. Like our AR meetings, this can be used to discuss the features and get input from the community. The worry with slack and adhoc discussions is that it gets buried in those blackholes and might not make it to the doc. But I guess we will have to do that regardless. 

One option is -- discuss in slack, sync meetings and capture those in the doc. 

Will add this point.

File: docs/proposals/VIP_TEMPLATE.md
Comment (manujose0) [reply docs/proposals/VIP_TEMPLATE.md:8]: let me explore the best way to capture this.

File: docs/proposals/VIP_TEMPLATE.md
Comment (sixpluszero) [inline docs/proposals/VIP_TEMPLATE.md:104]: nit: Wondering if this is a big scope design, do we want to add a session or subsession to describe how we can break it down into different phases?

File: docs/CONTRIBUTING.md
Comment (FelixGV) [inline docs/CONTRIBUTING.md:58]: Let's use the alias: http://slack.venicedb.org in case community inviter stops working well in the future, there'll be fewer links to tweak later on.

File: docs/dev_guide/design_doc.md
Comment (FelixGV) [reply docs/dev_guide/design_doc.md:62]: I guess I'm envisioning the following workflow:

1. The proposer submits a PR with the VIP in either DRAFT or DISCUSSION stage.
2. We do a first round of discussion on the PR. Depending on the outcome of the discussion, the status could move to ACCEPTED or REJECTED, or it could stay in DISCUSSION stage (e.g. if we agree tentatively on the broad strokes, but there are still action items to refine certain aspects). At the end of this, the PR gets merged, and at that point the VIP will appear in the directory of all historical VIPs.
3. If the merged VIP is still in DISCUSSION stage, then further PRs will be submitted to address the remaining action items.
4. If the VIP is ACCEPTED, then next stages are implementation, with eventually some code change PR also carrying an alteration of the VIP page to move the status to IMPLEMENTED.

End result is that you can look at the directory of VIPs and see all of them, both accepted and rejected, and even some which are still in discussion stage. Only the ones which are still in the first round of PR review would not show up in the directory.

LMK if that makes sense.

File: docs/dev_guide/design_doc.md
Comment (FelixGV) [reply docs/dev_guide/design_doc.md]: Note that our public slack history is archived and searchable via http://linen.venicedb.org (even though the slack is a free plan with limited retention...). So I think the visibility aspect is not too bad. Also, while it is fine to have meetings such as the contributor sync to go over these proposals, I think written chat is still less ephemeral than meetings. IDK which proportion of VIPs would benefit from the chat, maybe only the higher complexity ones... but in any case, I'm hoping to channel the desire to chat in a way that it gets recorded somewhere easy to find, hence the idea of per-VIP channels.

File: clients/da-vinci-client/src/test/java/com/linkedin/davinci/kafka/consumer/KafkaClusterBasedRecordThrottlerTest.java
Comment (ZacAttack) [inline clients/da-vinci-client/src/test/java/com/linkedin/davinci/kafka/consumer/KafkaClusterBasedRecordThrottlerTest.java:39]: Whats up with this code change?

Comment (adamxchen) [reply clients/da-vinci-client/src/test/java/com/linkedin/davinci/kafka/consumer/KafkaClusterBasedRecordThrottlerTest.java:39]: super strange behavior of Spotbug. Spotbug didn't detect it before but somehow discover this today 😂

Comment (ZacAttack) [reply clients/da-vinci-client/src/test/java/com/linkedin/davinci/kafka/consumer/KafkaClusterBasedRecordThrottlerTest.java:39]: ack.  Ok.

Comment (FelixGV) [reply clients/da-vinci-client/src/test/java/com/linkedin/davinci/kafka/consumer/KafkaClusterBasedRecordThrottlerTest.java:39]: Thanks for staying vigilant that managers don't start doing code changes, Zac ;P

File: docs/CONTRIBUTING.md
Comment (FelixGV) [inline docs/CONTRIBUTING.md:56]: Let's update this to Wednesday.

File: docs/dev_guide/recommended_development_workflow.md
Comment (FelixGV) [inline docs/dev_guide/recommended_development_workflow.md:12]: s/feature/functionality/

**Review by FelixGV — COMMENTED** [review]
A couple more minor comments.

File: docs/CONTRIBUTING.md
Comment (manujose0) [reply docs/CONTRIBUTING.md:56]: sounds good. Done.

File: docs/CONTRIBUTING.md
Comment (manujose0) [reply docs/CONTRIBUTING.md:58]: Good Idea. Done.

File: docs/dev_guide/recommended_development_workflow.md
Comment (manujose0) [reply docs/dev_guide/recommended_development_workflow.md:12]: Thanks for catching this in here as well.

File: docs/proposals/VIP_TEMPLATE.md
Comment (manujose0) [reply docs/proposals/VIP_TEMPLATE.md:104]: Good point. I was thinking the development milestone be covered by the implementation team realm docs but you are right it can be used to sub divide a complex project. Thanks for the suggestion. Added the same.

File: docs/dev_guide/design_doc.md
Comment (manujose0) [reply docs/dev_guide/design_doc.md]: So there are essentially 3 ways to discuss 1/ PR 2/ Slack with dedicated channel 3/ sync meetings. Updated.

File: docs/proposals/VIP_TEMPLATE.md
Comment (manujose0) [reply docs/proposals/VIP_TEMPLATE.md:8]: Figured out how to achieve this this spotless. Will do that in a separate PR as I need to do some more testing and will take time on this one.

File: docs/dev_guide/design_doc.md
Comment (manujose0) [reply docs/dev_guide/design_doc.md]: Sounds good. Updated the instructions as well.

**Review by eolivelli — CHANGES_REQUESTED** [review]
Awesome document.

Overall LGTM but I would add a section about 'Rejected altenatives'

File: docs/proposals/VIP_TEMPLATE.md
Comment (FelixGV) [inline docs/proposals/VIP_TEMPLATE.md:4]: Hmm, this seems inconsistent with the path below. It might work but may make the page structure confusing. Either we have a new top-level section for proposals, at the same level as the dev guides section, or else proposals are nested under dev guides.

File: docs/proposals/VIP_TEMPLATE.md
Comment (FelixGV) [inline docs/proposals/VIP_TEMPLATE.md:5]: Would probably be good to have a landing page for proposals which presents a table of contents of all VIPs, either auto-generated from the child pages, if that's possible, or manually curated (which may be better if we want to display extra info, e.g. the status, release version, or other details...

We could leave that for later since we don't have any VIPs to present yet but anyway...

cc @nisargthakkar in case you have opinions/advice on page structure.

**Comment by FelixGV** [comment]
> Awesome document.
> 
> 
> 
> Overall LGTM but I would add a section about 'Rejected altenatives'

Thanks for chiming in @eolivelli! The alternatives, including rejected ones, are currently tucked under the design section. Though maybe we should put them in a section of their own? IDK.

I guess each structure may be ideal for different phases of the design process. In the discussion phase, it may be best not to bias reviewers too much towards a specific alternative, so that the best one emerges based on its tradeoffs, rather than based solely on which section it's in. Whereas for accepted proposals, we may want to highlight more clearly the chosen versus rejected paths. IDK, what do you think?

File: docs/proposals/VIP_TEMPLATE.md
Comment (manujose0) [reply docs/proposals/VIP_TEMPLATE.md:4]: So Proposals are at the same level as the dev guide but looks like the parent in these sections points to "Community Guides". Will do that as proposals are part of how we interact with the community. Looking into that one, I figured we also have design guidelines pointing to RFC wiki on the contributing page. Will correct that one too.

File: docs/proposals/VIP_TEMPLATE.md
Comment (manujose0) [reply docs/proposals/VIP_TEMPLATE.md:5]: Yeah. Planning to do a landing page. Let's get the first one in hopefully @ZacAttack might soon publish something after this gets merged. 

Let me do a landing page anyway in anticipation that this will land soon.

**Review by eolivelli — APPROVED** [review]
Let's start with this model and then we can iterate.


I think that as a follow up step, as the community grows, we need to define the process that leads to "The proposal has been accepted by the Venice project".

In established OSS projects, like Pulsar, Kafka....there is a VOTE process.
This is not possible at the moment, as that would mean that there is a committee of people who have binding VOTEs.

File: docs/proposals/VIP_TEMPLATE.md
Comment (manujose0) [reply docs/proposals/VIP_TEMPLATE.md]: Does this look alright @FelixGV

**Review by FelixGV — APPROVED** [review]
Thanks a lot! Looking forward to seeing where this takes us!

**Pull request merged**
